### DISCUSSION TITLE
How to move Clover to a hard disk on a Linux system?

### DISCUSSION THREAD

**Discussion by alan-rpi** [CloverHackyColor/CloverBootloader#658] [Q&A] [answered]
First, thank you to author of Clover for making this software available. My searches on the internet for my question all seem to find Mac examples.

I've installed LinuxLite on a SSD in a PCIe socket on an old PC and can use Clover from a USB stick to boot into Clover and select the SSD. All fine :)

The PC also has a spinning HD - the original boot device. I've reformatted this with a 90MB boot/eps partition sda1 and the rest of the drive as a ext4 data partition. In LinuxLite I cannot see the boot/eps partition so I booted into a live Linux from a USB stick and in a terminal window I can now access it:

sudo mount /dev/sda1 /mnt/sda1 
sudo ls /dev/sda1
lost+found

I insert the Clover USB stick and can access that:
 ls media/linux/DOSUTILS
 Bootloaders   EFI        'System Volume Information'   ThirdParty
 BootSectors   rcScripts   themespkg

So, should I copy the whole contents of media/linux/DOSUTILS to /dev/sda1  or only some folders? And is there anything else to do - presumably I need to change the Clover config default boot disk to the SSD boot partition or has Clover remembered that somewhere?

I hope someone can answer this as the SSD has made a substantial difference and I would like to make a more permanent solution without booting requiring a USB stick. 

Alan

**Comment by Morbius01** [comment]
Guessing what you are trying to do : you would like to install Clover on your HDD and boot Linux Lite from it rather than using the USB Clover key. Is that correct ?

I managed to do this on my legacy system, not on a bootable HDD but on a bootable SATA SSD (all my OSs are installed on a non bootable PCIe NVMe SSD).

For this, I just cloned the Clover key to the SATA SSD, using the Linux _**ddrescue**_ utility.

To install ddrescue : 
`sudo apt install gddrescue ddrescue`

To clone the Clover key to the bootable drive : 
`ddrescue /dev/<source> /dev/<target> --verbose [--force]`
 \<source> : Clover key Id
\<target> : bootable drive (your HDD) Id

Does it work ?

**Reply by alan-rpi** [reply]
Many thanks for trying to help. 

> Guessing what you are trying to do : you would like to install Clover on your HDD and boot Linux Lite from it rather than using the USB Clover key. Is that correct ?
Spot on :)

I tried copying the Clover USB stick EFI folders to the HDD boot partition but it did not work. I think a problem is I reformatted the HDD and whilst I created a boot partition it is empty of course.    

ddrescue looks the tool. I would have to clone the boot partition on the SSD but then the UUID is the same and this has previously caused me problems. I can use GParted to generate new UUID but I gather that grub configuration file within the boot partition is full of references to the original SSD partition. Am I right?

I've decided that it might initially be easier to reinstall Linux on the HDD and then copy the EFI folder from the Clover USB stick to the HDD boot.

I'll report back

**Reply by Morbius01** [reply]
I just did it again, and it works like a charm !
And I didn't bother about any UUID issue...

Note that when I used ddrescue, I use the entire device (USB Clover key) as a source, not just the boot partition.
Meaning that, if the USB Clover key is _sdb_ and the boot partition is _sdb1_, and if the target SATA drive is _sdc_, I did :
`sudo ddrescue /dev/sdb /dev/sdc --verbose --force`
It's the only way which seems to work for me.
Doing this way will make the **sdb1** partition to be cloned to sdc including the _boot_ & _lba_ flags (essential !).

Any luck on your side ?

**Comment by alan-rpi** [comment]
Well my plan worked :) 

I installed Linux to the HDD. I had to remove the SSD as the installation wanted to format the swap file on the SSD - why?  
I then listed the HD boot partition:
sudo mkdir /mnt/sda1
sudo mount /dev/sda1 /mnt/sda1 	
sudo ls /mnt/sda1			
EFI  startup.nsh
and strangely this gives the same:
sudo ls /mnt/sda1/EFI		
EFI  startup.nsh

I then listed the Clover USB stick
ls /media/user/DOSUTILS/EFI/
BOOT  CLOVER

and copied this to the HDD boot  drive (actually I did it twice):
sudo cp -rf /media/user/DOSUTILS/EFI /mnt/sda1/EFI
sudo ls /mnt/sda1/EFI/EFI
BOOT  CLOVER  ubuntu
sudo cp -rf /media/user/DOSUTILS/EFI /mnt/sda1
sudo ls /mnt/sda1/EFI
BOOT  CLOVER  EFI  startup.nsh

I rebooted and accessed the bios with F2 and F8 and was presented with an extra boot option 4:
1. ubuntu (P1 <HDD id>)
2. ubuntu (P1 <HDD id>)
3. PI: <HDD id> (250GB)
4. UEFI OS (P1: <HDD id>)
Why there were 3 HDD ones to start with I don't know.
I chose 4 and got the Clover boot menu with 1 choice for the HDD.

I then powered off and reinserted the SSD and adapter in the PCIe slot and rebooted.
But now only choices 1 to 3. 
Rebooted again and the 4 choices were back -phew! 
So I set the boot priority to choice 4 and booted
Now I have the Clover menu with 2 choices: HDD and SSD
Clover seems to remember the choice so another boot defaulted to the SSD.

Only slight problem is when booting to the SSD after the Clover screen I get a Grub menu which currently has along timeout to default tp normal startup. Need to find how to shorten that.

Now I just need to boot many times to make sure that the 4th Clover option doesn't disappear and the PC boots to the HDD.

I then need to repeat this on a Windows 10 PC but I cannot afford the time for a complete fresh install of Win10 and can't mess it up as it's my wife's PC. For this PC I've already cloned the HDD to a SSD. 

=================
As per your suggestion to use ddrescue. You refer to the "USB Clover key" - do you mean "USB Clover stick"?

The problem is my USB stick is 1GB (it's old) but the HDD boot partition is just 200MB.
The stick contains the following 1 file and folders
.DS_Store
rcScripts
themespkg
Bootloaders
EFI
BootSectors
ThirdParty

I could make the HDD boot partition 1GB but reading your statement:

> Meaning that, if the USB Clover key is sdb and the boot partition is sdb1, and if the target SATA drive is sdc, I did :
> sudo ddrescue /dev/sdb /dev/sdc --verbose --force

suggests this will completely wipe the HDD

Perhaps we have slightly different use cases. Did you want to replicate a complete system whereas I just want a HDD bootable partition with Clover?

I much appreciate you taking the time to help @Morbius01 :)

Alan in the UK
Now midnight so calling it a day. Catch up tomorrow.

**Comment by alan-rpi** [comment, marked as answer]
Today the PC has booted fine many times into Clover and defaults to booting the SSD.

I booted Linux with a USB stick and set the Partition Names and Labels via Gparted - the partitions need to be unmounted, hence using a live usb. Names are xxxx_SSD. Now Clover shows xxxxxxxx Boot_SSD as the boot drive name.

I also changed the Grub menu timeout from 6? secs to 3 secs - just a few terminal commands - do a web search.
